website/content/docs/rfcs/inline-auth.mdx

```diff
@@ -0,0 +1,179 @@
+subsequently handled by), resulting in audit logs for both the authentication
+and the main request. Because the main request is modified to have
+authentication, audit logs correctly point to the inline authentication as if
+it occurred just prior to the request.
+
+Various storage entries in the token store and expiration manager are created
+in memory but not written. These are attached to the response from the inline
+authentication, to be attached to the inbound request. As this inline
+authentication response is not returned to the caller, it remains private.
+
+### Integration in Ecosystem
+
+Other projects in the ecosystem, such as OpenBao Agent, and k8s secrets operator
+will need to be updated to support this. Each auth method will need to be
+checked for support. This does not need to occur at the same time as the main
+implementation but should be saved as future work.
+
+Note that OpenBao Auto-Authing Proxy may not be able to function with this
+change; it will need to gracefully handle leases and will be likely used with
+many requests so the benefits are minimal.
+
+## Rationale and Alternatives
+
+Improving the efficacy of future horizontal scalability is important; this
+change helps with that. This improves overall performance as well, if inline
+authentication is used more broadly by reducing the total number of storage
+operations made by OpenBao.
+
+One alternative would be for applications to use longer-lived authentication;
+in many workloads, this is not possible.
+
+## Downsides
+
+The complexity of this change is rather small (~300LOC without tests).
+However, it does introduce a few new code paths as expected by a change of
+this nature.
+
+## Security Implications
+
+Authentication will still be performed; this is equivalent to a three-request
+operation:
+
+ 1. Authenticate to OpenBao.
+ 2. Issue actual request.
+ 3. Revoke token.
+
+However, the application does not see the token ever.
+
+## User/Developer Experience
+
+This change is backwards compatible with previous API versions, including to
+HashiCorp's Vault API client. As long as the operator can modify each
+request's headers or authentication values are sufficiently static, this
+should work with third-party clients, especially those that expect to talk
+to Agent for transparent authentication.
+
+## Unresolved Questions
+
+- Is the header format sufficiently defensive?
+- Will header-size limiting become an issue for certain request types?
+- Not clear is if this will affect quotas or other subsystems.
+- Not clear if this should have a separate `X-Vault-Inline-Auth-Namespace`
+  header to support cross-namespace requests/authentication.
+- Should `core.LocalUpdateUserFailedLoginInfo(...)` be skipped on inline
+  auth requests?
+
+## Related Issues
+
+- https://gitlab.com/gitlab-org/gitlab/-/issues/540889
+
+## Proof of Concept
+
+https://github.com/cipherboy/openbao/pull/new/inline-authentication
```

Comment:
The proof of concept had a bug, wherein token policies were not respected. I refactored this to behave better, using the cached TokenEntry value.

Reply:
It also now invalidates leases and errs out accordingly, with a simple test of this functionality (policy behavior & lease revocation).